ostd/src/mm/page/allocator.rs

```diff
@@ -282,3 +349,121 @@ pub(crate) fn init() {
     }
     PAGE_ALLOCATOR.call_once(|| SpinLock::new(allocator));
 }
+
+/// The bootstrapping phase page allocator.
+pub(crate) struct BootstrapFrameAllocator {
+    // memory region idx: The index for the global memory region indicates the
+    // current memory region in use, facilitating rapid boot page allocation.
+    mem_region_idx: usize,
+    // frame cursor: The cursor for the frame which is the next frame to be
+    // allocated.
+    frame_cursor: usize,
+}
+
+impl BootstrapFrameAllocator {
+    pub fn new() -> Self {
+        // Get the first frame for allocation
+        let mut first_idx = 0;
+        let mut first_frame = 0;
+        let regions = crate::boot::memory_regions();
+        for i in 0..regions.len() {
+            if regions[i].typ() == crate::boot::memory_region::MemoryRegionType::Usable {
+                // Make the memory region page-aligned, and skip if it is too small.
+                let start = regions[i].base().align_up(PAGE_SIZE) / PAGE_SIZE;
+                let end = regions[i]
+                    .base()
+                    .checked_add(regions[i].len())
+                    .unwrap()
+                    .align_down(PAGE_SIZE)
+                    / PAGE_SIZE;
+                log::debug!(
+                    "Found usable region, start:{:x}, end:{:x}",
+                    regions[i].base(),
+                    regions[i].base() + regions[i].len()
+                );
+                if end <= start {
+                    continue;
+                } else {
+                    first_idx = i;
+                    first_frame = start;
+                    break;
+                }
+            }
+        }
+        Self {
+            mem_region_idx: first_idx,
+            frame_cursor: first_frame,
+        }
+    }
+
+    pub fn get_frame_cursor(self) -> usize {
+        self.frame_cursor
+    }
+}
+
+impl PageAlloc for BootstrapFrameAllocator {
+    fn add_frame(&mut self, _start: usize, _end: usize) {
+        warn!("BootstrapFrameAllocator does not need to add frames");
+    }
+
+    fn alloc(&mut self, _layout: Layout) -> Option<Paddr> {
+        warn!("BootstrapFrameAllocator does not support to allocate memory described by range");
+        None
+    }
+
+    fn alloc_page(&mut self, _align: usize) -> Option<Paddr> {
+        let frame = self.frame_cursor;
+        debug!("allocating frame: {:#x}", frame * PAGE_SIZE,);
+        // Update idx and cursor
+        let regions = crate::boot::memory_regions();
+        self.frame_cursor += 1;
+        loop {
+            let region = regions[self.mem_region_idx];
```

Comment:
According to the definition, the memory region is a global static vector `Vec<MemoryRegion>` while `boot::memory_regions()` is its getter function, returning the vector's reference. 

If we need to cache `region`s' information, it is natural to copy it to a local vector, which is similar to how `region`s' information is originally stored and used. Therefore, whether cache the `region` in the bootstrap allocator's context or not will have a subtle impact on the function's performance. 

Besides, since the heap space is limited while bootstrapping, we should avoid using heap space in the bootstrap allocator, the minimal implementation of page allocator.